codecov_auth/views/okta_cloud.py

```diff
@@ -0,0 +1,200 @@
+    """The Codecov app page we redirect users to."""
+    return f"{settings.CODECOV_DASHBOARD_URL}/{service}/{org_username}"
+
+
+def get_oauth_redirect_url(org_username: str, service: str) -> str:
+    """The Okta callback URL for us to finish the authentication."""
+    return f"{settings.CODECOV_API_URL}/login/okta/callback"
+
+
+def get_okta_settings(organization: Owner) -> OktaSettings | None:
+    account: Account | None = organization.account
+    if account:
+        okta_settings: OktaSettings | None = account.okta_settings.first()
+        if okta_settings:
+            return okta_settings
+
+    log.warning(
+        "No Okta settings found",
+        extra={"org_owner_id": organization.ownerid},
+    )
+    return None
+
+
+class OktaCloudLoginView(OktaLoginMixin, View):
+    service = "okta_cloud"
+
+    def get(
+        self, request: HttpRequest, service: str, org_username: str
+    ) -> HttpResponse:
+        log_context: dict = {"service": service, "username": org_username}
+        if not request.session.get("current_owner_id"):
+            log.warning(
+                "User needs to be signed in before authenticating organization with Okta.",
+                extra=log_context,
+            )
+            return HttpResponse(status=403)
+
+        try:
+            organization: Owner = Owner.objects.get(
+                service=service, username=org_username
+            )
+        except Owner.DoesNotExist:
+            log.warning("The organization doesn't exist.", extra=log_context)
+            return HttpResponse(status=404)
+
+        okta_settings = get_okta_settings(organization)
+        if not okta_settings:
+            log.warning(
+                "Okta settings not found. Cannot sign into Okta", extra=log_context
+            )
+            return HttpResponse(status=404)
+
+        app_redirect_url = get_app_redirect_url(org_username, service)
+        oauth_redirect_url = get_oauth_redirect_url(org_username, service)
+
+        # User is already logged in, redirect them to the org page
+        if organization.account.id in request.session.get(
+            OKTA_SIGNED_IN_ACCOUNTS_SESSION_KEY, []
+        ):
+            return redirect(app_redirect_url)
+
+        # Otherwise start the process redirect them to the Issuer page to authenticate
+        else:
+            consent = self._redirect_to_consent(
+                iss=okta_settings.url,
+                client_id=okta_settings.client_id,
+                oauth_redirect_url=oauth_redirect_url,
+            )
+            request.session[OKTA_CURRENT_SESSION] = {
+                "org_ownerid": organization.ownerid,
+                "okta_settings_id": okta_settings.id,
+            }
+            return consent
+
+
+class OktaCloudCallbackView(OktaLoginMixin, View):
+    service = "okta_cloud"
+
+    def get(self, request: HttpRequest) -> HttpResponse:
+        current_okta_session: dict[str, int] = request.session.get(OKTA_CURRENT_SESSION)
```

Comment:
as you introduced pylance in part 1, would it make sense to use a stricter checked type here rather than `dict[str, int]` (which is also missing a `| None`)?

Reply:
Ah whoops good catch! I'll fix this.

> would it make sense to use a stricter checked type here rather than dict[str, int]

Unfortunately, `request.session` needs to be serializable (so can't just store the models directly). I considered creating a Pydantic model to dump back into a dict to go into a session (i.e. `okta_session.model_dump()`) but thought that this object might be too simple to need several rounds of conversion. Thoughts on whether it's worth going back to this?